src/main/scala/chisel3/util/MemInterface.scala

```diff
@@ -0,0 +1,337 @@
+    * @param size The desired size of the inner `SyncReadMem`
+    * @tparam T The data type of the memory element
+    * @param numRd The number of desired read ports, >= 0
+    * @param numWr The number of desired write ports, >= 0
+    * @param numRdWr The number of desired read/write ports, >= 0
+    * @param clock The clock to bind to each generated port, which may be different from the implicit clock
+    *
+    * @return A new `MemInterface` wire containing the control signals for each instantiated port
+    * @note This does *not* return the wrapper module itself, you must interact with it using the returned bundle
+    */
+  def withMask[T <: Data](
+    size:    BigInt,
+    tpe:     T,
+    numRd:   Int,
+    numWr:   Int,
+    numRdWr: Int,
+    clock:   Clock
+  )(
+    implicit evidence: T <:< Vec[_]
+  ): MemInterface[T] =
+    macro MemTransform.memInterface_withMaskClk[T]
+
+  /** @group SourceInfoTransformMacro */
+  def do_apply[T <: Data](
+    size:    BigInt,
+    tpe:     T,
+    numRd:   Int,
+    numWr:   Int,
+    numRdWr: Int
+  )(
+    implicit sourceInfo: SourceInfo
+  ): MemInterface[T] = memInterface_impl(size, tpe)(numRd, numWr, numRdWr, Builder.forcedClock)
+
+  /** @group SourceInfoTransformMacro */
+  def do_apply[T <: Data](
+    size:    BigInt,
+    tpe:     T,
+    numRd:   Int,
+    numWr:   Int,
+    numRdWr: Int,
+    clock:   Clock
+  )(
+    implicit sourceInfo: SourceInfo
+  ): MemInterface[T] = memInterface_impl(size, tpe)(numRd, numWr, numRdWr, clock)
+
+  /** @group SourceInfoTransformMacro */
+  def do_withMask[T <: Data](
+    size:    BigInt,
+    tpe:     T,
+    numRd:   Int,
+    numWr:   Int,
+    numRdWr: Int
+  )(
+    implicit sourceInfo: SourceInfo,
+    evidence:            T <:< Vec[_]
+  ): MemInterface[T] = memInterface_withMask_impl(size, tpe)(numRd, numWr, numRdWr, Builder.forcedClock)
+
+  /** @group SourceInfoTransformMacro */
+  def do_withMask[T <: Data](
+    size:    BigInt,
+    tpe:     T,
+    numRd:   Int,
+    numWr:   Int,
+    numRdWr: Int,
+    clock:   Clock
+  )(
+    implicit sourceInfo: SourceInfo,
+    evidence:            T <:< Vec[_]
+  ): MemInterface[T] = memInterface_withMask_impl(size, tpe)(numRd, numWr, numRdWr, clock)
+
+  /** @group SourceInfoTransformMacro */
+  private def memInterface_impl[T <: Data](
+    size:    BigInt,
+    tpe:     T
+  )(numRd:   Int,
+    numWr:   Int,
+    numRdWr: Int,
+    clock:   Clock
+  )(
+    implicit sourceInfo: SourceInfo
+  ): MemInterface[T] = {
+    val addrWidth = log2Up(size + 1)
+
+    val _wrappedMem = Module(new Module {
+      override def desiredName: String =
+        s"SyncReadMemWrapper_${MemInterface.portedness(numRd, numWr, numRdWr)}_${tpe.typeName}"
+
+      val io = IO(new MemInterface(tpe, addrWidth, numRd, numWr, numRdWr, false))
+
+      val _innerMem = SyncReadMem(size, tpe)
+
+      for (i <- 0 until numRd) {
+        io.rd(i).readValue := _innerMem.read(io.rd(i).addr, io.rd(i).enable, clock)
```

Comment:
You're passing a clock as an argument but also injecting a level of module hierarchy. The passed clock is not in the right module so this cannot be right and I'm shocked there is no error being thrown. Please figure out why this isn't erroring (because it should be) and then remove this clock argument. If we're going to pass a clock as an argument, you can only use it by driving ports of the wrapping Module, but probably should just be doing a `withClock` around the module instantiation (if we even want the Module).